packages/SwingSet/src/kernel/kernel.js

```diff
@@ -286,35 +274,35 @@ export default function buildKernel(
     queueToKref(vatAdminRootKref, 'meterCrossedThreshold', args, 'logFailure');
   }
 
-  function deductMeter(meterID, compute, firstTime) {
-    assert.typeof(compute, 'bigint');
-    const res = kernelKeeper.deductMeter(meterID, compute);
+  // TODO: instead of using a kernel-wide flag here, consider making each
+  // VatManager responsible for remembering if/when a KernelSyscallResult
+  // reports a non-'ok' status and therefore the vat is toast. Then the
+  // delivery handler could ask the manager (or vat-warehouse) afterwards for
+  // the sticky-fatal state. If we did that, we wouldn't need
+  // `vatFatalSyscall`. We'd still need a way for `requestTermination` to
+  // work, though.
```

Comment:
Oh, you know, we should consider making the `syscall.exit` handler just push an `terminate-vat` event onto the vat's own input-queue, and perform the teardown later. Sort of "please shut me down, eventually, when it's convenient". Then vat-manager or the translation layer doesn't need to remember anything. And if our scheduler's policy always drains the vat-output-queue before looking at the vat-input-queue, then this would automatically deliver all the previously-emitted work before shutdown, which might line up nicely with our plans around involuntary termination needing to not interfere with previous output, and might simplify the "vat is dead but queues are not" question.

I don't know enough about our use cases for this to know whether promptness is important to us. Maybe it'd be when a contract has finished transferring all its assets and obligations to a successor, and it self-destructs to free up chain capacity, and to prevent surprises when people try to talk to the old one by mistake? Seems like the vat would know that it's in this state, so it could maybe reject all attempts to start new work while waiting to be terminated. OTOH, it would be really annoying to have to add extra "am I active" checks in vat code to make up for the platform not being able to respond to a `syscall.exit` promptly. We might work around this by configuring the input scheduler to service `terminate-vat` events ahead of anything else (but still behind draining the output queue).